actions/notification_actions.jsx

```diff
@@ -17,121 +17,132 @@ import {stripMarkdown} from 'utils/markdown';
-        username = post.props.override_username;
-    } else if (msgProps.sender_name) {
-        username = msgProps.sender_name;
-    } else if (UserStore.hasProfile(post.user_id)) {
-        username = UserStore.getProfile(post.user_id).username;
-    }
-
-    let title = Utils.localizeMessage('channel_loader.posted', 'Posted');
-    if (!channel) {
-        title = msgProps.channel_display_name;
-        channel = {
-            name: msgProps.channel_name,
-            type: msgProps.channel_type,
-        };
-    } else if (channel.type === Constants.DM_CHANNEL) {
-        title = Utils.localizeMessage('notification.dm', 'Direct Message');
-    } else {
-        title = channel.display_name;
-    }
-
-    if (title === '') {
-        if (msgProps.channel_type === Constants.DM_CHANNEL) {
-            title = Utils.localizeMessage('notification.dm', 'Direct Message');
+export async function sendDesktopNotification(post, msgProps) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const currentUserId = getCurrentUserId(state);
+
+        if ((currentUserId === post.user_id && post.props.from_webhook !== 'true')) {
+            return;
+        }
+
+        if (isSystemMessage(post)) {
+            return;
+        }
+
+        let mentions = [];
+        if (msgProps.mentions) {
+            mentions = JSON.parse(msgProps.mentions);
+        }
+        const teamId = msgProps.team_id;
+
+        let channel = getChannel(state, post.channel_id);
+        const user = getCurrentUser(state);
+        const userStatus = getStatusForUserId(state, user.id);
+        const member = getMyChannelMember(state, post.channel_id);
+
+        if (isChannelMuted(member) || userStatus === UserStatuses.DND || userStatus === UserStatuses.OUT_OF_OFFICE) {
+            return;
+        }
+
+        let notifyLevel = member && member.notify_props ? member.notify_props.desktop : NotificationLevels.DEFAULT;
+        if (notifyLevel === NotificationLevels.DEFAULT) {
+            notifyLevel = user && user.notify_props ? user.notify_props.desktop : NotificationLevels.ALL;
+        }
+
+        if (notifyLevel === NotificationLevels.NONE) {
+            return;
+        } else if (notifyLevel === NotificationLevels.MENTION && mentions.indexOf(user.id) === -1 && msgProps.channel_type !== Constants.DM_CHANNEL) {
+            return;
+        }
+
+        const config = getConfig(state);
+        const userFromPost = getUser(state, post.user_id);
+        let username = '';
+        if (post.props.override_username && config.EnablePostUsernameOverride === 'true') {
+            username = post.props.override_username;
+        } else if (userFromPost) {
+            username = displayUsername(userFromPost, getTeammateNameDisplaySetting(state), false);
         } else {
+            const missingProfileResponse = await dispatch(getProfilesByUsernames([msgProps.sender_name]));
```

Comment:
1. `msgProps.sender_name` may not actually be the user's name here. It can also be the overridden name for an integration post, so in that case, we'd want to have `username = msgProps.sender_name` without going to the server. We should probably change the server to make sure that `sender_name` is always the sender's username since the client already handles the overridden name case a few lines above this
2. Could we maybe refactor this out to happen before the rest of this function? Awaiting in the middle of a big function like this could have some unexpected side effects like how `state` is incorrect after this because the store may have been updated before this.

Reply:
I'll see what I can come up with

Reply:
I'm still not fond of the await in the middle of that, but the rest of the changes look good. I'll make a note about possibly deprecating the sender_name field on that event